S3 notification event trigger Java-based AWS Lambda function (#2399)
aws-samples/serverless-patterns: s3-lambda-resizing-java/README.md

## Requirements

* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have enough permissions to make necessary AWS service calls and manage AWS resources.
* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
* [AWS Serverless Application Model](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) (AWS SAM) installed

## Deployment Instructions

-  Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
   ``` 
   git clone https://github.com/aws-samples/serverless-patterns
   ```
- Change directory to the pattern directory:
    ```
    cd s3-lambda-resizing-java
    ```
- From the command line, use AWS SAM to build and deploy the AWS resources for the pattern as specified in the `template.yml` file:
    ```
    sam build && sam deploy --guided
    ```
- During the prompts:
  * Enter a stack name
  * Enter the desired AWS Region

>   * Allow SAM CLI to create IAM roles with the required permissions.

Reviewer: ```suggestion
  * Allow SAM CLI to create IAM roles with the required permissions.
      * Enter names for your source and destination S3 buckets. Make sure these are unique as S3 bucket names share a global namespace.
```

> # AWS Amazon S3 to AWS Lambda — Create a Java-based Lambda function that resizes images uploaded to S3

Reviewer 2: ```suggestion
# Resizing images uploaded to Amazon S3 with AWS Lambda in Java
```

The SAM template deploys:
 - A Java-based Lambda function:
   - Powered by [Amazon Q Developer](https://aws.amazon.com/q/developer/)
   - [**NEW**] [Lambda SnapStart enabled for ARM64](https://aws.amazon.com/about-aws/whats-new/2024/07/aws-lambda-snapstart-java-functions-arm64-architecture/) based architecture
 - A source S3 bucket 
 - A destination S3 bucket
 - The IAM resources required to run the application. 

The Lambda function consumes `ObjectCreated` events from an Amazon S3 bucket. 
The Lambda code checks the uploaded file is an image and creates a lower resolution version of the image in the destination bucket.

Learn more about this pattern at Serverless Land Patterns: [https://serverlessland.com/patterns/s3-lambda](https://serverlessland.com/patterns/s3-lambda)

Important: this application uses various AWS services, and there are costs associated with these services after the Free Tier usage.
Please see the [AWS Pricing page](https://aws.amazon.com/pricing/) for details. 
You are responsible for any AWS costs incurred. No warranty is implied in this example.

## Requirements

> 

Reviewer 2: Add requirement for Java21

  * Allow SAM CLI to create IAM roles with the required permissions.

Once you have run above once and saved arguments to a configuration file (`samconfig.toml`), 
you can use `sam deploy` in future to use these defaults.

Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.

## How it works

* Use the AWS CLI upload an image to S3
* If the object is a .jpeg, the code creates an image of 800x600 resolution and saves it to the destination bucket.

## Testing

Run the following S3 CLI command to upload an image to the S3 bucket. 
Note, you must edit the {SourceBucketName} and {DestinationBucketName} placeholder with the name of the S3 Bucket. 
This is provided in the stack outputs.

```bash
aws s3 cp './data/white_dog.jpeg'  s3://{SourceBucketName}
```

Run the following command to check that a new version of the image has been created in the destination bucket.

```bash
aws s3 ls s3://{DestinationBucketName}
```

## Running JUnit 5 test
To run the JUnit test `AppTest.java`, you need to make few changes:

> - Modify the [events/s3_event.json](./ResizerFunction/src/test/resources) file to reflect the source bucket name and object key. Replace these values in the placeholders.

Reviewer 2: Can we default the object key to white_dog.jepg?

- Replace the destination bucket name placeholder in the surefire plugin's environment variable placeholder in the `pom.xml` file.
- Enable the test [`AppTest.java`](./ResizerFunction/src/test/java/resizer/AppTest.java) by removing the `@Disabled` annotation.

> - Run `mvn clean test`

Reviewer 2: Add changing directory to ./ResizerFunction
